### PR TITLE
VCB-199 String implementation for dcb

### PR DESCRIPTION
**Pull request by stefenz**

### CODE REVIEW

File: src/assembler/encoder.ts [hunk 1]
```diff
@@ -237,19 +237,53 @@ function writeDataInstruction(
...
   let bytes: Byte[] = []
   let alignment: number = 0
-  const values = instruction.options.map((x) => +x)
+  let optionsContainAString: boolean = false
+  let values: number[] = []
+  instruction.options.forEach((x) => {
+    if (x.startsWith('"')) {
+      optionsContainAString = true
+      x = x.substring(0, x.length - 1).substring(1)
+      const charArray: string[] = x.split('')
```
Comment (m-dot-wigger) [inline src/assembler/encoder.ts:246]: instead of making char array. use charCodeAt with i as param and i+1 for escapted "

**Review by m-dot-wigger — CHANGES_REQUESTED** [review]
Please add a test in the encoder to make sure that it is interpreted correctly

File: src/assembler/parser/index.ts [hunk 2]
```diff
@@ -5,8 +5,9 @@ import { ITextMatch, ITextParseRule, parseText } from './text'
 const SYMBOL = `[a-z_]+[a-z0-9_]*|\\|[a-z0-9._ ]+\\|`
 const VALUE = `[0-9a-z#]+`
 const SPACE_OR_TAB = `[ \\t]`
+const STRING = `("(?:[^'"\n]|'"?)*")`
 
-const OPTION = `(?:[0-9a-z#=_]|[\\[{]${SPACE_OR_TAB}*|${SPACE_OR_TAB}*[\\]}]|${SPACE_OR_TAB}*-${SPACE_OR_TAB}*)+`
+const OPTION = `((?:[0-9a-z#=_]|[\\[{]${SPACE_OR_TAB}*|${SPACE_OR_TAB}*[\\]}]|${SPACE_OR_TAB}*-${SPACE_OR_TAB}*)+|${STRING})`
```
Comment (m-dot-wigger) [inline src/assembler/parser/index.ts:10]: Please make regex groups non-matching if not used within on match function

File: test/assembler/encoder.test.ts [hunk 3]
```diff
@@ -58,6 +58,36 @@ describe('encode', function () {
...
+      symbols: {},
+      areas: [
+        {
+          type: AreaType.Data,
+          isReadOnly: true,
+          name: '|.data|',
+          instructions: [
+            {
+              name: 'DCB',
+              options: ['"TestString"', `"TestString with Escape'""`],
```
Comment (m-dot-wigger) [inline test/assembler/encoder.test.ts:72]: perhaps add additional escape in middle of string

File: test/assembler/encoder.test.ts [hunk 4]
```diff
@@ -58,6 +58,36 @@ describe('encode', function () {
...
+    }
+    const file = encode(code)
+    expect(Object.keys(file.sections).length).toBe(1)
+    expect(getSection(file, '|.data|').offset).toBe(0)
+    expect(getSection(file, '|.data|').size).toBe(33)
+    expect(file.content.length).toBe(33)
+    expect(file.content[0].value).toBe(84)
+    expect(file.content[1].value).toBe(101)
+    expect(file.content[9].value).toBe(103)
+    expect(file.content[19].value).toBe(103)
```
Comment (m-dot-wigger) [inline test/assembler/encoder.test.ts:87]: test  file.content[10] start of second string

File: test/assembler/encoder.test.ts [hunk 5]
```diff
@@ -58,6 +58,36 @@ describe('encode', function () {
...
+    expect(file.content[25].value).toBe(32)
+    expect(file.content[32].value).toBe(34)
```
Comment (m-dot-wigger) [inline test/assembler/encoder.test.ts:89]: check file.content[31] to be e and not single quote

File: src/assembler/encoder.ts [hunk 6]
```diff
@@ -296,19 +296,49 @@ function writeDataInstruction(
...
+  let optionsContainAString: boolean = false
+  let values: number[] = []
+  instruction.options.forEach((x) => {
+    if (x.startsWith('"')) {
+      optionsContainAString = true
+      x = x.substring(0, x.length - 1).substring(1)
+      for (let i = 0; i < x.length; i++) {
+        if (i + 1 < x.length) {
+          // Special case when the string contains an escaped double quote (i.e. "Example'"")
+          if (x.charCodeAt(i) === 39 && x.charCodeAt(i + 1) === 34) {
```
Comment (m-dot-wigger) [inline src/assembler/encoder.ts:308]: as on #53 use charAt and compare with '"' and"'" 
so there are no magic numbers in code

**Review by m-dot-wigger — CHANGES_REQUESTED** [review]

File: src/assembler/encoder.ts [hunk 7]
```diff
@@ -296,19 +296,49 @@ function writeDataInstruction(
   }
   let bytes: Byte[] = []
   let alignment: number = 0
-  const values = instruction.options.map((x) => +x)
+  let optionsContainAString: boolean = false
+  let values: number[] = []
+  instruction.options.forEach((x) => {
+    if (x.startsWith('"')) {
+      optionsContainAString = true
+      x = x.substring(0, x.length - 1).substring(1)
```
[the remark this answers is not in the document]
Comment (stefenz) [reply src/assembler/encoder.ts]: Done

File: test/assembler/encoder.test.ts (hunk 5 above)
Comment (stefenz) [reply test/assembler/encoder.test.ts]: Done

File: test/assembler/encoder.test.ts (hunk 4 above)
Comment (stefenz) [reply test/assembler/encoder.test.ts:88]: Done

File: test/assembler/encoder.test.ts (hunk 3 above)
Comment (stefenz) [reply test/assembler/encoder.test.ts]: Done

File: src/assembler/parser/index.ts (hunk 2 above)
Comment (stefenz) [reply src/assembler/parser/index.ts]: Done

File: src/assembler/encoder.ts (hunk 6 above)
Comment (stefenz) [reply src/assembler/encoder.ts]: Done

**Review by m-dot-wigger — APPROVED** [review]
Thank you die the changes

**Pull request merged**